ui: add parameterized dispatch interface

nicholascioli opened the pull request:
Closes #9313 

This commit adds a new interface for dispatching parameteried jobs, if
the user has the right permissions. The UI can be accessed by viewing a
parameterized job and clicking on the "Dispatch Job" button located in
the "Job Launches" section.

(Note: This commit adds one dependency `title-case` for pretty printing of 
meta names. This can be removed if necessary.)

Given the following configuration:

```json
...
"Meta": {
    "traefik_service_name": "traefik",
    "another_optional_item": "item-default",
    "server_name": "server-name-1"
  },
"ParameterizedJob": {
  "MetaOptional": [
    "traefik_service_name",
    "another_optional_item"
  ],
  "MetaRequired": [
    "server_name"
  ],
  "Payload": "optional"
},
...
```

The following screenshots show the generated interface.

![Parameterized Job Overview](https://user-images.githubusercontent.com/1240491/120281314-c915fe00-c286-11eb-95fc-5383e515aea2.png)

![Creating the dispatched job](https://user-images.githubusercontent.com/1240491/120281349-d632ed00-c286-11eb-99b0-76016bedd569.png)

![Showing an error when applicable](https://user-images.githubusercontent.com/1240491/120281381-e0ed8200-c286-11eb-852b-4371c0af8040.png)

![Dispatched job with supplied meta](https://user-images.githubusercontent.com/1240491/120281421-eb0f8080-c286-11eb-8113-dfa51365c4e2.png)

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/nomad?pullRequest=10675)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/nomad?pullRequest=10675)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/nomad?pullRequest=10675).</sub>

nicholascioli commented:
Is that failing test something that I can address? I don't know what is meant by `Resource not accessible by integration`...

tgross commented:
Wow, thanks @nicholascioli! Just a heads up that it might take us a little bit to get to reviewing this one as we're in the middle of bringing on new UI folks who will want to have a say here.

I think the "resource not available by integration" is a GitHub Actions auth issue... it looks like that specific test needs a GitHub token but we don't want to expose those to PRs from the community. So don't worry about that for now.

nicholascioli commented:
I cleaned up a few things that seemed inconsistent in both my changes and the interface. Updated screenshots are attached below:

![Screenshot 2021-06-08 at 00-02-50 Dispatch new traefik - Nomad](https://user-images.githubusercontent.com/1240491/121121533-bb63f980-c7ed-11eb-9280-41d10a95cb76.png)

![Screenshot 2021-06-08 at 00-03-29 Job traefik dispatch-1623125004-a86da8ee - Nomad](https://user-images.githubusercontent.com/1240491/121121537-bef78080-c7ed-11eb-9590-f1e42132c905.png)

lgfa29 commented on ui/app/templates/components/job-dispatch.hbs:
```diff
@@ -0,0 +1,60 @@
+      <div class="columns">
+        <div class="column is-4">
+          <h2 class="title is-4">{{meta.title}} {{#if meta.isRequired}}*{{/if}}</h2>
+          <input
+            id="{{meta.name}}"
+            style="width: 100%;"
```
These `input` fields are not styled like other fields. We normally just use `class="input"` and avoid in-line styles. 

Here's an existing field for reference: https://github.com/hashicorp/nomad/blob/main/ui/app/templates/settings/tokens.hbs#L26

lgfa29 commented on ui/app/templates/components/job-dispatch.hbs:
```diff
@@ -0,0 +1,60 @@
+            oninput={{action "updateParamValue" meta.name}}
+            placeholder={{meta.placeholder}}
+            required={{meta.isRequired}} >
+          <small>{{#if meta.isRequired}}Required{{else}}Optional{{/if}} Meta Param ({{ meta.name }})</small>
```
The original meta param name could be highlighted a bit more, maybe using something like 

```html
<span class="badge is-light is-faded">
  <code>{{ meta.name }}</code>
</span>
```

lgfa29 commented on ui/app/templates/components/job-dispatch.hbs:
```diff
@@ -0,0 +1,60 @@
+    </div>
+
+    <div class="bumper-top">
+      <button class="button is-success" type="button" onclick={{perform this.submit}}>Dispatch</button>
```
```suggestion
      <button class="button is-primary" type="button" onclick={{perform this.submit}}>Dispatch</button>
```

lgfa29 commented on ui/app/templates/components/job-dispatch.hbs:
```diff
@@ -0,0 +1,60 @@
+            style="width: 100%;"
+            value={{this.paramValues.[meta.name]}}
+            oninput={{action "updateParamValue" meta.name}}
+            placeholder={{meta.placeholder}}
```
Using a placeholder for the default value forces users to re-type the default meta value.

lgfa29 commented on ui/app/templates/components/job-page/parts/children.hbs:
```diff
@@ -1,5 +1,8 @@
 <div class="boxed-section-head">
   Job Launches
+  {{#if (can "dispatch job")}}
+    <LinkTo @route="jobs.job.dispatch" data-test-run-job class="button is-primary pull-right">Dispatch Job</LinkTo>
```
This button seems misaligned with the `collapse` button above it.

lgfa29 commented on ui/app/templates/components/job-dispatch.hbs:
```diff
@@ -0,0 +1,60 @@
+          <input
+            id="{{meta.name}}"
+            style="width: 100%;"
+            value={{this.paramValues.[meta.name]}}
```
Is the `.[` correct here?

lgfa29 reviewed (changes requested):
Thank you for this wonderful feature PR @nicholascioli!

I flagged some issues that we found while reviewing, but most are small style changes. The main blockers is the lack of `bas64` encoding of the payload and the use of `originalTarget`, which causes the form to fail on non-Firefox browsers.

You don't have to worry about making any of these changes, @JBhagat841 and I will work on them, but if you feel like tackling some of them before we do feel free to go ahead 😄 

Thank you again for the contribution!

lgfa29 commented on ui/app/models/job-dispatch.js:
```diff
@@ -0,0 +1,10 @@
+  @attr() evalCreateIndex;
+  @attr() evalID;
+  @attr() dispatchedJobID;
+}
```
Ops...it is necessary, otherwise the dispatch action fails.

JBhagat841 commented on ui/app/abilities/job.js:
```diff
@@ -29,4 +32,9 @@ export default class Job extends AbstractAbility {
   }
+
+  @computed('rulesForNamespace.@each.capabilities')
+  get policiesSupportDispatching() {
```
I think we'll also need a test and some refactoring of existing tests to account for `dispatch-job` in `abilities/job-test`

JBhagat841 commented on ui/tests/unit/adapters/job-test.js:
```diff
@@ -387,6 +389,27 @@ module('Unit | Adapter | Job', function(hooks) {
+    const job = await this.initializeWithJob();
+    job.set('parameterized', true);
+
+    const payload = "I'm a payload 🙂";
```
Would it be helpful if we explain why we're specifically making sure to use an emoji here? Maybe editing the `payload` variable name.

lgfa29 replied:
Text-based tests usually involve adding some non-ASCII characters (like the exec alloc tests). I think it's fairly common in the code base.

JBhagat841 commented on ui/app/components/job-dispatch.js:
```diff
@@ -0,0 +1,138 @@
+        x =>
+          new MetaField({
+            name: x,
+            required: required,
```
nit when the object key and value are the same you can get away with,
```
{
name: x,
required,
otherKey: otherValue
}
```

JBhagat841 commented on ui/app/components/job-dispatch.js:
```diff
@@ -0,0 +1,138 @@
+      );
+
+    // Fetch the different types of parameters.
+    const required = mapper(this.args.job.parameterizedDetails.MetaRequired || [], true);
```
No change needed, JS ES6 Knowledge Transfer Comment:

We're using `||` in the event that we have `null` or `undefined` here?

You should checkout the nullish coalescing operator `??` which specifically handles the cases of `null` and `undefined` but will return the left hand operand if its a falsy value like `0` or `''`.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator

lgfa29 replied:
Ah good to know, thanks! In this case it doesn't seem to matter, since falsey values are not expected for our `Meta*` attributes.

lgfa29 replied:
I actually changed the function signature to:
```js
const mapper = (values = [], required) {
  ...
}
```

Seems more readable and reliable. Thanks for making double check 😄

JBhagat841 commented on ui/app/components/job-dispatch.js:
```diff
@@ -0,0 +1,138 @@
+    this.router.transitionTo('jobs.job');
+  }
+
+  @(task(function*() {
+    // Try to create the dispatch.
+    try {
+      let paramValues = {};
+      this.metaFields.forEach(m => (paramValues[m.name] = m.value));
+      const dispatch = yield this.args.job.dispatch(paramValues, this.payload);
+
+      // Navigate to the newly created instance.
+      this.router.transitionTo('jobs.job', dispatch.DispatchedJobID);
+    } catch (err) {
+      const error = messageFromAdapterError(err) || 'Could not dispatch job';
+      this.errors.pushObject(error);
+      this.scrollToError();
+    }
+  }).drop())
+  onDispatched;
```
Question:  Is this syntax up to date for new Ember Octane components?

This is new to me.

http://ember-concurrency.com/docs/task-function-syntax/

JBhagat841 commented on ui/app/utils/encode.js:
```diff
@@ -0,0 +1,22 @@
+    string = '';
+  }
+
+  let encoded = new TextEncoderLite('utf-8').encode(string);
```
nit:  no re-assignment so this can be a `const`

JBhagat841 reviewed (commented):
LGTM, will take another look in the afternoon after we chat.

JBhagat841 reviewed (approved):
LGTM
